include/seahorn/Expr/ExprRewriter.hh

```diff
@@ -0,0 +1,177 @@
+#pragma once
+#include "seahorn/Expr/Expr.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprRewriteRule.hh"
+#include "seahorn/Support/SeaDebug.h"
+#include "seahorn/Support/SeaLog.hh"
+
+namespace expr {
+using namespace seahorn;
+using namespace mem;
+
+namespace utils {
+bool shouldCache(Expr e);
+} // end of namespace utils
+
+/** rewrite hybrid memory expr in ITE structure
+ * loadMem: Expr tree that represents value loaded from current memory state
+ * ptr: address Expr tree
+ **/
+Expr rewriteHybridLoadMemExpr(Expr loadMem, Expr ptr, AddrRangeMap &arm);
+
+struct RewriteFrame {
+  Expr m_exp;     // the Expr node (subtree) under rewrite
+  size_t m_depth; // number of levels to rewrite from this node
+  size_t m_i;     // up to m_i th children have been rewritten
+  RewriteFrame(Expr exp, size_t depth, size_t i)
+      : m_exp(exp), m_depth(depth), m_i(i) {}
+};
+
+using RewriteFrameVector = std::vector<RewriteFrame>;
+
+class ExprRewriterConfig {
+  /* apply rewrite rules */
+protected:
+  ExprFactory &m_efac;    // for making expr
+  DagVisitCache &m_cache; // for accessing cached rewrite results
+
+public:
+  ExprRewriterConfig(ExprFactory &efac, DagVisitCache &cache, AddrRangeMap &arm)
+      : m_efac(efac), m_cache(cache) {}
+
+  rewrite_result applyRewriteRules(Expr exp);
+
+  bool shouldRewrite(Expr exp);
+};
+
+/* example config for ITE */
+class ITECompRewriteConfig : public ExprRewriterConfig {
+private:
+  ITERewriteRule m_iteRule;
+  CompareRewriteRule m_compRule;
+  BoolOpRewriteRule m_boolRule;
+  ArrayRewriteRule m_arrayRule;
+  ArithmeticRule m_arithRule;
+
+public:
+  ITECompRewriteConfig(ExprFactory &efac, DagVisitCache &cache,
+                       AddrRangeMap &arm)
+      : m_iteRule(efac, cache), m_compRule(efac, cache),
+        m_boolRule(efac, cache), m_arrayRule(efac, cache, arm),
```

Comment:
Looks like you only use ARM for `select` only. Maybe I did not fully get the usage of the ARM. Can Arm be used for other pointer expressions?

Reply:
we can perhaps use it during the initial rewrite from `store`s to `ITE`s